File: Sources/CLI/Connection.swift
import Foundation
import AgentComputerShared

/// Unix domain socket client for communicating with the daemon.
enum Connection {
    
    /// Send a request to the daemon and receive a response.
    /// Auto-starts the daemon if it's not running.
    static func send(_ request: Request, verbose: Bool = false) throws -> Response {
        let socketPath = daemonSocketPath.path
        
        // Try to connect, auto-start daemon if needed
        let fd = try connectOrStartDaemon(socketPath: socketPath, verbose: verbose)
        defer { close(fd) }
        
        // Encode request as JSON line
        let encoder = JSONEncoder()
        let data = try encoder.encode(request)
        guard var jsonLine = String(data: data, encoding: .utf8) else {
            throw ConnectionError.encodingFailed
        }
        jsonLine += "\n"
        
        if verbose {
            FileHandle.standardError.write(Data("[verbose] Sending: \(jsonLine.trimmingCharacters(in: .newlines))\n".utf8))
        }
        
        // Send
        let sendData = Array(jsonLine.utf8)
        let written = Darwin.write(fd, sendData, sendData.count)
        guard written == sendData.count else {
            throw ConnectionError.writeFailed
        }
        
        // Read response line
        let responseLine = try readLine(fd: fd)
        
        if verbose {
            FileHandle.standardError.write(Data("[verbose] Received: \(responseLine)\n".utf8))
        }
        
        // Decode response
        guard let responseData = responseLine.data(using: .utf8) else {
            throw ConnectionError.decodingFailed
        }
        let decoder = JSONDecoder()
        let response = try decoder.decode(Response.self, from: responseData)
        return response
    }
    
    /// Connect to existing socket, or start daemon and wait for it.
    private static func connectOrStartDaemon(socketPath: String, verbose: Bool) throws -> Int32 {
        // Try connecting first
        if let fd = tryConnect(socketPath: socketPath) {
            return fd
        }
        
        // Socket not available — start daemon
        FileHandle.standardError.write(Data("Starting agent-computer daemon...\n".utf8))
        try spawnDaemon(verbose: verbose)
        
        // Poll for socket availability (100ms intervals, 5s timeout)
        let deadline = Date().addingTimeInterval(5.0)
        while Date() < deadline {
            if let fd = tryConnect(socketPath: socketPath) {
                return fd
            }
            Thread.sleep(forTimeInterval: 0.1)
        }
        
        throw ConnectionError.daemonStartTimeout
    }
    
    /// Try to connect to Unix socket. Returns fd on success, nil on failure.
    private static func tryConnect(socketPath: String) -> Int32? {
        guard FileManager.default.fileExists(atPath: socketPath) else { return nil }
        
        let fd = socket(AF_UNIX, SOCK_STREAM, 0)
        guard fd >= 0 else { return nil }
        
        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)
        
        let pathBytes = socketPath.utf8CString
        guard pathBytes.count <= MemoryLayout.size(ofValue: addr.sun_path) else {
            close(fd)
            return nil
        }
        
        withUnsafeMutablePointer(to: &addr.sun_path) { ptr in
            ptr.withMemoryRebound(to: CChar.self, capacity: pathBytes.count) { dest in
                pathBytes.withUnsafeBufferPointer { src in
                    _ = memcpy(dest, src.baseAddress!, pathBytes.count)
                }
            }
        }
        
        let addrLen = socklen_t(MemoryLayout<sockaddr_un>.size)
        let result = withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                connect(fd, sockPtr, addrLen)
            }
        }
        
        if result < 0 {
            close(fd)
            return nil
        }
        
        return fd
    }
    
    /// Spawn the daemon as a background process.
    private static func spawnDaemon(verbose: Bool) throws {
        // Find the daemon binary relative to the CLI binary
        let cliBinaryPath = CommandInfo.executableURL
        let binDir = cliBinaryPath.deletingLastPathComponent()
        let daemonPath = binDir.appendingPathComponent("agent-computer-daemon")
        
        // Fallback: try PATH
        let finalPath: URL
        if FileManager.default.isExecutableFile(atPath: daemonPath.path) {
            finalPath = daemonPath
        } else {
            // Try finding in PATH
            let whichProcess = Process()
            let whichPipe = Pipe()
            whichProcess.executableURL = URL(fileURLWithPath: "/usr/bin/which")
            whichProcess.arguments = ["agent-computer-daemon"]
            whichProcess.standardOutput = whichPipe
            whichProcess.standardError = FileHandle.nullDevice
            try? whichProcess.run()
            whichProcess.waitUntilExit()
            
            let whichData = whichPipe.fileHandleForReading.readDataToEndOfFile()
            let whichOutput = String(data: whichData, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            
            if !whichOutput.isEmpty && FileManager.default.isExecutableFile(atPath: whichOutput) {
                finalPath = URL(fileURLWithPath: whichOutput)
            } else {
                throw ConnectionError.daemonNotFound
            }
        }
        
        if verbose {
            FileHandle.standardError.write(Data("[verbose] Spawning daemon: \(finalPath.path)\n".utf8))
        }
        
        let process = Process()
        process.executableURL = finalPath
        process.standardOutput = FileHandle.nullDevice
        process.standardError = FileHandle.nullDevice
        
        // Ensure socket directory exists
        try FileManager.default.createDirectory(at: daemonSocketDir, withIntermediateDirectories: true)
        
        try process.run()
        // Don't wait — daemon runs in background
    }
    
    /// Read a newline-terminated line from a file descriptor.
    private static func readLine(fd: Int32) throws -> String {
        var buffer = [UInt8]()
        var byte: UInt8 = 0
        let maxSize = 10 * 1024 * 1024 // 10MB max response
        
        while buffer.count < maxSize {
            let n = Darwin.read(fd, &byte, 1)
            if n <= 0 {
                if buffer.isEmpty {
                    throw ConnectionError.connectionClosed
                }
                break
            }
            if byte == UInt8(ascii: "\n") {
                break
            }
            buffer.append(byte)
        }
        
        guard let line = String(bytes: buffer, encoding: .utf8) else {
            throw ConnectionError.decodingFailed
        }
        return line
    }
}

// MARK: - Helpers

private enum CommandInfo {
    static var executableURL: URL {
        URL(fileURLWithPath: ProcessInfo.processInfo.arguments[0])
    }
}

// MARK: - Errors

enum ConnectionError: Error, CustomStringConvertible {
    case daemonNotFound
    case daemonStartTimeout
    case connectionFailed
    case connectionClosed
    case writeFailed
    case encodingFailed
    case decodingFailed
    
    var description: String {
        switch self {
        case .daemonNotFound:
            return "Could not find agent-computer-daemon binary. Make sure it's in the same directory as agent-computer or on your PATH."
        case .daemonStartTimeout:
            return "Daemon did not start within 5 seconds. Check if another instance is running or if there are permission issues."
        case .connectionFailed:
            return "Failed to connect to daemon socket."
        case .connectionClosed:
            return "Daemon closed the connection unexpectedly."
        case .writeFailed:
            return "Failed to send data to daemon."
        case .encodingFailed:
            return "Failed to encode request as JSON."
        case .decodingFailed:
            return "Failed to decode response from daemon."
        }
    }
}

File: crates/cli/src/connection.rs
use agent_computer_shared::protocol::{Request, Response};
use agent_computer_shared::types::daemon_socket_path;

use std::io::{BufRead, BufReader, Write};
use std::os::unix::net::UnixStream;
use std::path::PathBuf;
use std::time::Duration;

// MARK: - Connection Errors

#[derive(Debug)]
pub enum ConnectionError {
    DaemonNotFound,
    DaemonStartTimeout,
    ConnectionFailed(String),
    ConnectionClosed,
    WriteFailed(String),
    EncodingFailed(String),
    DecodingFailed(String),
}

impl std::fmt::Display for ConnectionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::DaemonNotFound => write!(
                f,
                "Could not find agent-computer-daemon binary. Make sure it's in the same directory as agent-computer or on your PATH."
            ),
            Self::DaemonStartTimeout => write!(
                f,
                "Daemon did not start within 5 seconds. Check if another instance is running or if there are permission issues."
            ),
            Self::ConnectionFailed(e) => write!(f, "Failed to connect to daemon socket: {}", e),
            Self::ConnectionClosed => {
                write!(f, "Daemon closed the connection unexpectedly.")
            }
            Self::WriteFailed(e) => write!(f, "Failed to send data to daemon: {}", e),
            Self::EncodingFailed(e) => write!(f, "Failed to encode request as JSON: {}", e),
            Self::DecodingFailed(e) => {
                write!(f, "Failed to decode response from daemon: {}", e)
            }
        }
    }
}

impl std::error::Error for ConnectionError {}

// MARK: - Send

/// Send a request to the daemon and receive a response.
/// Auto-starts the daemon if it's not running.
pub fn send(request: &Request, verbose: bool) -> Result<Response, ConnectionError> {
    let socket_path = daemon_socket_path();

    // Try to connect, auto-start daemon if needed
    let stream = connect_or_start_daemon(&socket_path, verbose)?;
    let mut stream_write = stream;

    // Encode request as JSON line
    let json = serde_json::to_string(request).map_err(|e| ConnectionError::EncodingFailed(e.to_string()))?;
    let mut json_line = json.clone();
    json_line.push('\n');

    if verbose {
        eprintln!("[verbose] Sending: {}", json);
    }

    // Send
    stream_write
        .write_all(json_line.as_bytes())
        .map_err(|e| ConnectionError::WriteFailed(e.to_string()))?;
    stream_write
        .flush()
        .map_err(|e| ConnectionError::WriteFailed(e.to_string()))?;

    // Read response line
    let mut reader = BufReader::new(&stream_write);
    let mut response_line = String::new();
    reader
        .read_line(&mut response_line)
        .map_err(|e| ConnectionError::DecodingFailed(e.to_string()))?;

    if response_line.is_empty() {
        return Err(ConnectionError::ConnectionClosed);
    }

    if verbose {
        eprintln!("[verbose] Received: {}", response_line.trim());
    }

    // Decode response
    let response: Response = serde_json::from_str(response_line.trim())
        .map_err(|e| ConnectionError::DecodingFailed(e.to_string()))?;

    Ok(response)
}

// MARK: - Connect or Start Daemon

fn connect_or_start_daemon(
    socket_path: &PathBuf,
    verbose: bool,
) -> Result<UnixStream, ConnectionError> {
    // Try connecting first
    if let Some(stream) = try_connect(socket_path) {
        return Ok(stream);
    }

    // Socket not available — start daemon
    eprintln!("Starting agent-computer daemon...");
    spawn_daemon(verbose)?;

    // Poll for socket availability (100ms intervals, 5s timeout)
    let deadline = std::time::Instant::now() + Duration::from_secs(5);
    while std::time::Instant::now() < deadline {
        if let Some(stream) = try_connect(socket_path) {
            return Ok(stream);
        }
        std::thread::sleep(Duration::from_millis(100));
    }

    Err(ConnectionError::DaemonStartTimeout)
}

/// Try to connect to Unix socket. Returns stream on success, None on failure.
fn try_connect(socket_path: &PathBuf) -> Option<UnixStream> {
    if !socket_path.exists() {
        return None;
    }

    UnixStream::connect(socket_path).ok()
}

/// Spawn the daemon as a background process.
fn spawn_daemon(verbose: bool) -> Result<(), ConnectionError> {
    // Find the daemon binary relative to the CLI binary
    let cli_path = std::env::current_exe().ok();
    let daemon_path = cli_path
        .as_ref()
        .and_then(|p| p.parent())
        .map(|dir| dir.join("agent-computer-daemon"));

    let final_path = if let Some(ref path) = daemon_path {
        if path.is_file() {
            path.clone()
        } else {
            find_in_path("agent-computer-daemon")?
        }
    } else {
        find_in_path("agent-computer-daemon")?
    };

    if verbose {
        eprintln!("[verbose] Spawning daemon: {}", final_path.display());
    }

    // Ensure socket directory exists
    let socket_dir = agent_computer_shared::types::daemon_socket_dir();
    let _ = std::fs::create_dir_all(&socket_dir);

    // Spawn daemon as background process
    std::process::Command::new(&final_path)
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .spawn()
        .map_err(|e| {
            ConnectionError::ConnectionFailed(format!(
                "Failed to spawn daemon at {}: {}",
                final_path.display(),
                e
            ))
        })?;

    Ok(())
}

/// Find a binary in PATH.
fn find_in_path(name: &str) -> Result<PathBuf, ConnectionError> {
    let output = std::process::Command::new("which")
        .arg(name)
        .output()
        .map_err(|_| ConnectionError::DaemonNotFound)?;

    if output.status.success() {
        let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
        if !path.is_empty() {
            return Ok(PathBuf::from(path));
        }
    }

    Err(ConnectionError::DaemonNotFound)
}
